src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs

```diff
@@ -90,68 +78,54 @@ public void InvokeCompletionCallback(uint sniErrorCode)
         /// Allocate space for data
         /// </summary>
         /// <param name="capacity">Length of byte array to be allocated</param>
-        public void Allocate(int capacity)
+        public void Allocate(int capacity, bool reserveMuxHeader)
         {
-            if (_data != null && _data.Length < capacity)
+            SNIPacketFlags flags = reserveMuxHeader ? SNIPacketFlags.MuxHeaderReserved : SNIPacketFlags.None;
+            int headerCapacity = reserveMuxHeader ? SNISMUXHeader.HEADER_LENGTH : 0;
+            int totalCapacity = headerCapacity + capacity;
+            if (_data != null)
             {
-                if (_isBufferFromArrayPool)
+                if (_data.Length < totalCapacity)
                 {
-                    ArrayPool<byte>.Shared.Return(_data);
+                    Array.Clear(_data, 0, _header + _length);
+                    if ((_flags & SNIPacketFlags.ArrayFromPool) == SNIPacketFlags.ArrayFromPool)
+                    {
+                        ArrayPool<byte>.Shared.Return(_data, clearArray: false);
+                        _flags &= ~SNIPacketFlags.ArrayFromPool;
+                    }
+                    _data = null;
+                }
+                else
+                {
+                    // if the current array is big enough and rented keep it
+                    flags |= (_flags & SNIPacketFlags.ArrayFromPool);
                 }
-                _data = null;
             }
 
             if (_data == null)
```

Comment:
_data is null at this point, you can get rid of the check,.

Reply:
Done.